File: a2/Search1.py
import sys
import time

def buildIndex(filename):
	try:
		f = open(filename, "r")
	except:
		print "Cannot open file %s! Maybe you made a typo :(" % filename
		sys.exit()

	courses = {}

	for line in f.readlines():
		info = line.split("            ")
		courses[info[0]] = info[1] 
	
	return courses

def search(index, keyword):
	for target in index.keys():
		if keyword.lower() in index[target].lower():
			print target, " : ", index[target]

def resetTimer():
  global START
  START = time.time()

def readTimerMilliseconds():
  t = time.time() - START
  return (t * 1000)


#----------------------------------------------
if len(sys.argv[1:]) < 2:
	print "Usage <file> <keyword>"
	sys.exit()

filename = sys.argv[1]
keyword = sys.argv[2]

courses = buildIndex(filename)
resetTimer()
search(courses, keyword)
print "Took %f milliseconds to search for keyword %s" % (readTimerMilliseconds(), keyword)

File: a2/Search1.go
package main

import (
	"fmt"
	"os"
	"bufio"
	"strings"
	"time"
)

type Index map[string]string

func buildIndex(filename string) Index {
	f, err := os.Open(filename)
	if err != nil {
		fmt.Printf("Cannot open file %s \n", filename)
		os.Exit(0)
	}

	courses := make(Index)

	reader := bufio.NewReader(f)
	for line, err := reader.ReadString('\n');
       	err == nil;
       	line, err = reader.ReadString('\n') {
       	info := strings.Split(line, "            ")
       	courses[info[0]] = info[1]
    }

	return courses
}

func search(index Index, keyword string) {
	for key, value := range index {
		if(strings.Contains(strings.ToLower(value), strings.ToLower(keyword)) == true) {
			fmt.Println(key, " : ", value)
		}
	}
}

func main() {
	if len(os.Args) < 3 {
		fmt.Println("Usage <filename> <keyword>")
		os.Exit(0)
	}

	filename, keyword := os.Args[1], os.Args[2]
	courses := buildIndex(filename)
	start := time.Now()
	search(courses, keyword)
	duration := time.Now().Sub(start)*1000
	fmt.Println("Took ", duration, " milliseconds to search for keyword ", keyword)
}